In [73]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
#1). cat and numerical data   #smile for pharma industries
#2). The analysis u need to perform on Bank data set
#3). Analytics vidhay hackthon will pick 

In [74]:
path=r'C:\Users\manish\Datafile\Visadataset.csv'
visa_df=pd.read_csv(path)
visa_df.head()


case_id continent education_of_employee has_job_experience  \
0  EZYV01      Asia           High School                  N   
1  EZYV02      Asia              Master's                  Y   
2  EZYV03      Asia            Bachelor's                  N   
3  EZYV04      Asia            Bachelor's                  N   
4  EZYV05    Africa              Master's                  Y   

  requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                     N            14513         2007                 West   
1                     N             2412         2002            Northeast   
2                     Y            44444         2008                 West   
3                     N               98         1897                 West   
4                     N             1082         2005                South   

   prevailing_wage unit_of_wage full_time_position case_status  
0         592.2029         Hour                  Y      Denied  
1       83425.6500         Year                  Y   Certified  
2      122996.8600         Year                  Y      Denied  
3       83434.0300         Year                  Y      Denied  
4      149907.3900         Year                  Y   Certified

**Divide into categorical and numerical columns**

In [71]:
dict(visa_df.dtypes)

{'case_id': dtype('O'),
 'continent': dtype('O'),
 'education_of_employee': dtype('O'),
 'has_job_experience': dtype('O'),
 'requires_job_training': dtype('O'),
 'no_of_employees': dtype('int64'),
 'yr_of_estab': dtype('int64'),
 'region_of_employment': dtype('O'),
 'prevailing_wage': dtype('float64'),
 'unit_of_wage': dtype('O'),
 'full_time_position': dtype('O'),
 'case_status': dtype('O')}

In [72]:
types=dict(visa_df.dtypes)
cat_col=[key for key,value in types.items() if value=='object']
num_col=[key for key,value in types.items() if value!='object']

In [7]:
#shortcut method
cat_columns=visa_df.select_dtypes(include='object').columns
num_columns=visa_df.select_dtypes(exclude='object').columns
cat_columns,num_columns

(Index(['case_id', 'continent', 'education_of_employee', 'has_job_experience',
        'requires_job_training', 'region_of_employment', 'unit_of_wage',
        'full_time_position', 'case_status'],
       dtype='object'),
 Index(['no_of_employees', 'yr_of_estab', 'prevailing_wage'], dtype='object'))

**How  to selcet a specific columns**

In [9]:
visa_df['continent']

0          Asia
1          Asia
2          Asia
3          Asia
4        Africa
          ...  
25475      Asia
25476      Asia
25477      Asia
25478      Asia
25479      Asia
Name: continent, Length: 25480, dtype: object

In [10]:
visa_df[['continent']]

continent
0          Asia
1          Asia
2          Asia
3          Asia
4        Africa
...         ...
25475      Asia
25476      Asia
25477      Asia
25478      Asia
25479      Asia

[25480 rows x 1 columns]

In [11]:
visa_df.continent

0          Asia
1          Asia
2          Asia
3          Asia
4        Africa
          ...  
25475      Asia
25476      Asia
25477      Asia
25478      Asia
25479      Asia
Name: continent, Length: 25480, dtype: object

In [12]:
visa_df.loc[:,['continent']]

continent
0          Asia
1          Asia
2          Asia
3          Asia
4        Africa
...         ...
25475      Asia
25476      Asia
25477      Asia
25478      Asia
25479      Asia

[25480 rows x 1 columns]

In [13]:
# visa_df['continent']
# visa_df.loc[:,['continent']]
# visa_df.continent
# visa_df.loc[:,['continent']]


**How to read multiple columns in same time**
- whenever we see multiple means recall the list
- keep the columns in list
- pass the list


In [15]:
visa_df.loc[:,['continent','case_status']] 

continent case_status
0          Asia      Denied
1          Asia   Certified
2          Asia      Denied
3          Asia      Denied
4        Africa   Certified
...         ...         ...
25475      Asia   Certified
25476      Asia   Certified
25477      Asia   Certified
25478      Asia   Certified
25479      Asia   Certified

[25480 rows x 2 columns]

$Task-1$

**We want to find the  num of  applicants continent wise
- Ther is an Asia continent is there
- How many applicants are from Asia out of 25480

**Unique-method**
- uniqw will provide how many unique items ar ethere in a particular columns

In [18]:
visa_df['continent'].unique()

array(['Asia', 'Africa', 'North America', 'Europe', 'South America',
       'Oceania'], dtype=object)

In [19]:
#for counting
len(visa_df['continent'].unique())

6

**N-uniqe-method**

- gives count of unique items

In [21]:
visa_df['continent'].nunique()

6

In [22]:
# step-1: read the data with specific column
# step-2: when we apply the cond python will give true and false
# step-3: in order to read drive only true data dataframe_name[con]
con=visa_df['continent']=='Asia'  #just like if we give cond 10==10 it will give true just lik this
visa_df[con]
#len(visa_df[con])

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
5         EZYV06      Asia              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
5                         N             2339         2012   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
5                    South       78252.1400         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
2          Denied  
3          Denied  
5       Certified  
...           ...  
25475   Certified  
25476   Certified  
25477   Certified  
25478   Certified  
25479   Certified  

[16861 rows x 12 columns]

In [23]:
# con=visa_df['continent']=='Asia'  
# visa_df[con]

# con=visa_df['continent']=='Africa'  
# visa_df[con]

# con=visa_df['continent']=='North America'  
# visa_df[con]

# con=visa_df['continent']=='Europe'  
# visa_df[con]

# con=visa_df['continent']=='South America'  
# visa_df[con]

# con=visa_df['continent']=='Oceania'  
# visa_df[con]

In [24]:
for i in visa_df['continent'].unique():
    print(i)

Asia
Africa
North America
Europe
South America
Oceania


In [25]:
for i in visa_df['continent'].unique():
    con=visa_df['continent']==i  
    count=len(visa_df[con])
    print(f'The number of application from {i} is : {count}')

The number of application from Asia is : 16861
The number of application from Africa is : 551
The number of application from North America is : 3292
The number of application from Europe is : 3732
The number of application from South America is : 852
The number of application from Oceania is : 192


$task-2$

- Now instead of print count, take empty list
- Save the values in a empty list
- we already has unique continent list
- we also having count list
- create a Datframe

In [27]:
continent_list=[]
for i in visa_df['continent'].unique():
    continent_list.append(i)
continent_list
name_cont=pd.DataFrame(continent_list)
name_cont

0
0           Asia
1         Africa
2  North America
3         Europe
4  South America
5        Oceania

In [28]:
continents=visa_df['continent'].unique()
count=[]
for i in visa_df['continent'].unique():
    con=visa_df['continent']==i  
    count.append(len(visa_df[con]))
    
cols=['continets','Number of application']
pd.DataFrame(zip(continents,count),columns=cols)

continets  Number of application
0           Asia                  16861
1         Africa                    551
2  North America                   3292
3         Europe                   3732
4  South America                    852
5        Oceania                    192

**values counts**

In [30]:
visa_df['continent'].value_counts()

continent
Asia             16861
Europe            3732
North America     3292
South America      852
Africa             551
Oceania            192
Name: count, dtype: int64

In [31]:
visa_df['region_of_employment'].value_counts()

region_of_employment
Northeast    7195
South        7017
West         6586
Midwest      4307
Island        375
Name: count, dtype: int64

In [32]:
visa_df.dtypes

case_id                   object
continent                 object
education_of_employee     object
has_job_experience        object
requires_job_training     object
no_of_employees            int64
yr_of_estab                int64
region_of_employment      object
prevailing_wage          float64
unit_of_wage              object
full_time_position        object
case_status               object
dtype: object

In [33]:
case_ids=visa_df['case_id'].unique()
box=[]
for i in visa_df['case_id'].unique():
    con=visa_df['case_id']==i
    box.append(len(visa_df[con]))
pd.DataFrame(zip(case_ids,box),columns=['Case_ids','No.of_ids'])



Case_ids  No.of_ids
0         EZYV01          1
1         EZYV02          1
2         EZYV03          1
3         EZYV04          1
4         EZYV05          1
...          ...        ...
25475  EZYV25476          1
25476  EZYV25477          1
25477  EZYV25478          1
25478  EZYV25479          1
25479  EZYV25480          1

[25480 rows x 2 columns]

In [34]:
visa_df['region_of_employment'].unique()


array(['West', 'Northeast', 'South', 'Midwest', 'Island'], dtype=object)

In [35]:
employement=visa_df['region_of_employment'].unique()
count=[]
for i in visa_df['region_of_employment'].unique():
    con=visa_df['region_of_employment']==i  
    count.append(len(visa_df[con]))
pd.DataFrame(zip(employement,count),columns=['Number_region','Number  of emplyyent'])

    

Number_region  Number  of emplyyent
0          West                  6586
1     Northeast                  7195
2         South                  7017
3       Midwest                  4307
4        Island                   375

In [36]:
employee=visa_df['requires_job_training'].unique()
count=[]
for i in visa_df['requires_job_training'].unique():
    con=visa_df['requires_job_training']==i
    count.append(len(visa_df[con]))
count
pd.DataFrame(zip(employee,count),columns=['require_training','No._of_employye'])


require_training  No._of_employye
0                N            22525
1                Y             2955

In [37]:
years=visa_df['yr_of_estab'].unique()
count=[]
for i in visa_df['yr_of_estab'].unique():
    cond=visa_df['yr_of_estab']==i
    count.append(len(visa_df[cond]))
count
pd.DataFrame(zip(years,count),columns=['Years','No'])    

Years    No
0     2007   994
1     2002   435
2     2008   674
3     1897    37
4     2005  1051
..     ...   ...
194   1820     6
195   1809     9
196   1891    18
197   1830     6
198   1842     6

[199 rows x 2 columns]

In [38]:
vages=visa_df['unit_of_wage'].unique()
countt=[]
for i in visa_df['unit_of_wage'].unique():
    cond=visa_df['unit_of_wage']==i
    countt.append(len(visa_df[cond]))
countt
pd.DataFrame(zip(vages,count))   

0    1
0   Hour  994
1   Year  435
2   Week  674
3  Month   37

In [39]:
position_types=visa_df['full_time_position'].unique()
count_box=[]
for i in visa_df['full_time_position'].unique():
    cond=visa_df['full_time_position']==i
    count_box.append(len(visa_df[cond]))
pd.DataFrame(zip(position_types,count_box),columns=['Position_type','No.Required'])    

Position_type  No.Required
0             Y        22773
1             N         2707

In [40]:
case_study=visa_df['case_status'].unique()
box=[]
for i in visa_df['case_status'].unique():
    con=visa_df['case_status']==i
    box.append(len(visa_df[con]))
pd.DataFrame(zip(case_study,box),columns=['case_study','No.of employee'])    

case_study  No.of employee
0     Denied            8462
1  Certified           17018

In [41]:
len(visa_df['no_of_employees'].unique())

7105

In [42]:
employee=[no_ for no_ in visa_df['no_of_employees'].unique()]

count=[]
for i in visa_df['no_of_employees'].unique():
    con=visa_df['no_of_employees']==i
    ll=len(visa_df[con])
    count.append(ll)
pd.DataFrame(zip(employee,count),columns=['no_of_employee','Number'])


no_of_employee  Number
0              14513       1
1               2412       4
2              44444       1
3                 98       4
4               1082       9
...              ...     ...
7100           28199       1
7101           72892       1
7102           24335       1
7103           48785       1
7104           40224       1

[7105 rows x 2 columns]

In [43]:
wages=visa_df['prevailing_wage'].unique()
box=[]
for i in visa_df['prevailing_wage'].unique():
    con=visa_df['prevailing_wage']==i
    box.append(len(visa_df[con]))
pd.DataFrame(zip(wages,box),columns=['wages','values_items'])
    

wages  values_items
0         592.2029             1
1       83425.6500             1
2      122996.8600             1
3       83434.0300             1
4      149907.3900             2
...            ...           ...
25449   77092.5700             1
25450  279174.7900             1
25451  146298.8500             1
25452   86154.7700             1
25453   70876.9100             1

[25454 rows x 2 columns]

In [75]:
path=r'C:\Users\manish\Datafile\Visadataset.csv'
visa_df=pd.read_csv(path)
visa_df.head()


case_id continent education_of_employee has_job_experience  \
0  EZYV01      Asia           High School                  N   
1  EZYV02      Asia              Master's                  Y   
2  EZYV03      Asia            Bachelor's                  N   
3  EZYV04      Asia            Bachelor's                  N   
4  EZYV05    Africa              Master's                  Y   

  requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                     N            14513         2007                 West   
1                     N             2412         2002            Northeast   
2                     Y            44444         2008                 West   
3                     N               98         1897                 West   
4                     N             1082         2005                South   

   prevailing_wage unit_of_wage full_time_position case_status  
0         592.2029         Hour                  Y      Denied  
1       83425.6500         Year                  Y   Certified  
2      122996.8600         Year                  Y      Denied  
3       83434.0300         Year                  Y      Denied  
4      149907.3900         Year                  Y   Certified

In [6]:
def func(x, y=1): 
    return x * y 
print(func(5, 2))


10


In [8]:
def foo(a, b):
    a = b
    b = a
x = 10
y = 20
foo(x, y)
print(x, y)

10 20


In [24]:
def outer():
    x = 10
    def inner():
        nonlocal x
        x = 20
    inner()
    return x
print(outer())


20


In [44]:
def add():
    global a
    a=10
    b=20
    add=a+b
    return(b,add)
b,add=add()

In [50]:
def addition():
    NUMBER11=2000
    print("number1:",NUMBER11)
    print("number2:",NUMBER22)
    add=NUMBER111+NUMBER22
addition()

number1: 2000


NameError: name 'NUMBER22' is not defined

In [52]:
def greet1():
    print("===calling greet1 function=======")
    greet2()
    print("good morning")
def greet2():
    print("===calling greet2 function=======")
    print("good night")
greet2()

===calling greet2 function=======
good night


In [59]:
def greet1():
    print("Hello Sir")
    greet2()
    print("good morning")
def greet2():
    print("Thanks")
    print("See You")
greet2()


Thanks
See You


In [63]:
def mul():
    a=100
    b=700
    multiplication=a*b
    return(a,multiplication)
multiplication,a=mul()
print(multiplication,a)

100 70000


In [67]:
x = 50
def func():
   global x
   print('x is', x)
   x = 2
   print(x)
func()
print(x)


x is 50
2
2


In [71]:
def say(message, times = 1):
    print(message * times)
say('Hello')
say('World', 5)


Hello
WorldWorldWorldWorldWorld


In [83]:
def C2F(c):
    return c * 9/5 + 32
print C2F(100)
print C2F(0)

SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)? (734924806.py, line 3)

In [81]:
c=100
f=c*9/5+32
f

212.0

In [87]:
def power(x, y=2):
    r = 1
    for i in range(y):
       r = r * x
    return r
print(power(3))
print(power(3, 3))

9
27


In [89]:
def f(x, y, z): return x + y + z
f(2, 30, 400)

432

In [91]:
i=0
def change(i):
   i=i+1
   return i
change(1)
print(i)


0


In [99]:
def  change(i = 1, j = 2):
    i = i + j
    j = j + 1
    print(i, j)
change(j = 1, i = 2)


3 2


In [145]:
def add(a,b):
    c=a+b
    return c  
c=add(12,13)
print(c)

25


In [117]:
def no_return():
    pass  # No return statement

result = no_return()
print(result)  

None


In [147]:
def add(a, b):
    return a+5, b+5

result = add(3, 2)
print(result)

(8, 7)


In [149]:
def outer():
    x = 10
    def inner():
        print(x)
    inner()

outer()

10
